src/execution_plan/execution_plan.c

```diff
@@ -310,14 +325,37 @@ static void _ExecutionPlan_ProcessQueryGraph(ExecutionPlan *plan, QueryGraph *qg
 		}
 
 		if(cartesianProduct) {
-			// Add this traversal chain as a child under the Cartesian Product.
+			// We have multiple disjoint traversal chains.
+			// Add each chain as a child under the Cartesian Product.
 			ExecutionPlan_AddOp(cartesianProduct, root);
+		} else if(raxSize(bound_vars) > 0) {
+			/* We've built the only necessary traversal chain and have previously-bound variables.
+			 * Check to see if all previously-bound variables are bound in the traversal chain.
+			 * If they are, the traversal ops will form the new root and the previous ops will be pushed below.
+			 * Otherwise, the streams are disjoint and will be joined under a Cartesian product. */
+			rax *traversal_vars = raxNew();
+			ExecutionPlan_BoundVariables(root, traversal_vars); // Collect all variables in the traversal chain.
+
+			if(!_AllVariablesResolved(traversal_vars, bound_vars)) {
+				// Not all bound variables are represented in the traversal ops; the streams are disjoint. Example:
+				// MATCH (a) WITH AVG(a.age) AS avg MATCH (b) WHERE b.age < avg RETURN b
+				cartesianProduct = NewCartesianProductOp(plan);
+				_ExecutionPlan_UpdateRoot(plan, cartesianProduct);
+				ExecutionPlan_AddOp(cartesianProduct, root);
+			} else {
+				// All previously-bound variables are represented in the traversal ops, we have a single stream. Example:
+				// MATCH (a) WITH a MATCH (a)-[]->(b) RETURN b
+				_ExecutionPlan_UpdateRoot(plan, root);
+			}
+			raxFree(traversal_vars);
```

Comment:
I'm not sure I follow,
Running the example `MATCH (a) WITH AVG(a.age) AS avg MATCH (b) WHERE b.age < avg RETURN b` in Neo
doesn't uses a `cartesian product`, I can imagine the following:
`[SCAN a] -> [PROJECTION avg] -> [SCAN b]- > [FILTER b.age < avg] -> [PROJECTION b]`

the `[SCAN b]- > [FILTER b.age < avg]` part will operate on the same `avg` value for all bs in the graph until `[SCAN b]` is depleted in which case a new `avg` value is provided and the process repeat itself, execution terminates when `[PROJECTION avg]` is depleted.